src/helper/bitmap.js

```diff
@@ -145,7 +145,133 @@ const trim = function (raster) {
+            } else {
+                lastLeftMatchedColor = false;
+            }
+        }
+        if (x < imageData.width - 1) {
+            if (matchesColor_(x + 1, y, imageData, oldColor)) {
+                if (!lastRightMatchedColor) {
+                    stack.push([x + 1, y]);
+                    lastRightMatchedColor = true;
+                }
+            } else {
+                lastRightMatchedColor = false;
+            }
+        }
+    }
+};
+
+/**
+ * Flood fill beginning at the given point
+ * @param {!int} x The x coordinate on the context at which to begin
+ * @param {!int} y The y coordinate on the context at which to begin
+ * @param {!HTMLCanvas2DContext} context The context in which to draw
+ */
+const floodFill = function (x, y, context) {
+    const oldColor = getColor_(x, y, context);
+    context.fillRect(x, y, 1, 1);
+    const newColor = getColor_(x, y, context);
+    const imageData = context.getImageData(0, 0, context.canvas.width, context.canvas.height);
+    if (matchesColor_(x, y, imageData, oldColor)) { // no-op
+        return;
+    }
+    colorPixel_(x, y, imageData, newColor); // Restore old color to avoid affecting result
+    const stack = [[x, y]];
+    while (stack.length) {
+        const pop = stack.pop();
+        floodFillInternal_(pop[0], pop[1], imageData, newColor, oldColor, stack);
+    }
+    context.putImageData(imageData, 0, 0);
+};
+
+/**
+ * @param {!paper.Shape.Rectangle} rect The rectangle to draw to the canvas
+ * @param {!HTMLCanvas2DContext} context The context in which to draw
+ */
+const drawRect = function (rect, context) {
+    // No rotation component to matrix
+    if (rect.matrix.b === 0 && rect.matrix.c === 0) {
+        const width = rect.size.width * rect.matrix.a;
+        const height = rect.size.height * rect.matrix.d;
+        context.fillRect(
+            ~~(rect.matrix.tx - (width / 2)),
+            ~~(rect.matrix.ty - (height / 2)),
+            ~~width,
+            ~~height);
+        return;
+    }
+    const startPoint = rect.matrix.transform(new paper.Point(-rect.size.width / 2, -rect.size.height / 2));
+    const widthPoint = rect.matrix.transform(new paper.Point(rect.size.width / 2, -rect.size.height / 2));
+    const heightPoint = rect.matrix.transform(new paper.Point(-rect.size.width / 2, rect.size.height / 2));
+    const endPoint = rect.matrix.transform(new paper.Point(rect.size.width / 2, rect.size.height / 2));
+    const center = rect.matrix.transform(new paper.Point());
+    forEachLinePoint(startPoint, widthPoint, (x, y) => {
+        context.fillRect(x, y, 1, 1);
+    });
+    forEachLinePoint(startPoint, heightPoint, (x, y) => {
+        context.fillRect(x, y, 1, 1);
+    });
+    forEachLinePoint(endPoint, widthPoint, (x, y) => {
+        context.fillRect(x, y, 1, 1);
+    });
+    forEachLinePoint(endPoint, heightPoint, (x, y) => {
+        context.fillRect(x, y, 1, 1);
+    });
+    floodFill(~~center.x, ~~center.y, context);
```

Comment:
Could this helper be simplified by using the `rect.matrix` to transform the context and use `fillRect` always, instead of only in the special case of no rotation? Looks like paper has a utility for that kind of thing, `matrix.applyToContext`. I did some quick testing and this seems to work, including rotation/scaling:
```
const drawRect = function (rect, context) {
    context.save();
    rect.matrix.applyToContext(context);
    context.fillRect(
        ~~(-rect.size.width / 2), // Since rect matrix transforms to center
        ~~(-rect.size.width / 2),
        ~~rect.size.width,
        ~~rect.size.height);
    context.restore();
    return;
t / 2),
        ~~width,
        ~~height);
    context.restore();
    return;
}
```

Reply:
The problem with all of the canvas functions is that they create aliasing:
![image](https://user-images.githubusercontent.com/2855464/41119490-ed262bca-6a60-11e8-8638-0ee9cd5bdc93.png)
Otherwise paper.js's built in rasterize() function would make everything very easy!

The bitmap tools in Scratch 2.0 don't create aliasing, and that makes them much easier to work with with the fill tool and for making pixel art.